docs/docsite/rst/lookup_guide.rst

```diff
@@ -0,0 +1,159 @@
+Lookups and writes
+==================
+
+Lookups in Ansible kind of imply a read-only/non-destructive operation. Being run in templating, the expected result is that they *return* a value, and check mode has no effect on lookups. But there is no actual restriction on what a lookup can do, and some lookups do perform writes or otherwise change some state. For example, the ``password`` :ref:`lookup <ansible_collections.ansible.builtin.password_lookup>` writes a generated password to a file, to act as a sort of cache, and the ``pipe`` :ref:`lookup <ansible_collections.ansible.builtin.pipe_lookup>` runs an arbitrary shell command so it could easy write or change state.
+
+Writes in Vault
+---------------
+
+Operations that perform writes in Vault are not limited to the obvious ones such as writing a secret value, creating a policy, or enabling a new auth method.
+
+Any operation that creates a token for example, such as any login operation, is also a write; tokens use storage in Vault and having too many active tokens is a common cause of performance problems.
+
+Additionally, some values in Vault can only be "read" at the moment of their creation, and so the only way to retrieve such a value, is to get it as a response from the "write" that created it. A common example is AppRole secret IDs.
+
+The way this relates to Ansible and this collection, is that we may have lookup plugins that either unintuitively perform writes (like ``vault_login``), or appear inappropriate to exist as lookups in the first place, like the planned ``vault_write`` lookup.
+
+The reason for this is that the we often consider these operations to be logical "read" operations, like performing a login, and want to use their results in other expressions.
+
+Something like ``vault_write`` does not always fit that description, because you could use it in a way that is clearly an explicit write, for example you could create a new policy with the lookup. But there are times it may be appropriate to use it in lookup semantics, like when "retrieving" (really creating) a new secret ID for an approle.
+
+When considering built-in support for auth methods, any auth method other than ``token`` or ``none`` makes every lookup, even ``vault_read``, into something that's changing state and performing a write within Vault. This actually applies to many modules too, even when using check mode.
+
+How to reason about when to use lookups
+---------------------------------------
+
+Because there is potential for writes in any lookup, it is very important to carefully consider when you are using a lookup vs. a module/other plugin. Check mode has no effect on lookups, so there is potential to perform many writes within your check mode run, but maybe sometimes you want that, for example if you're performing a ``vault_login`` via lookup to retrieve a token to use in your module calls, you may want that to still happen in check mode so that your module calls can properly check the things they need to.
```

Comment:
Ahhhh, I see what you mean by "check mode has no effect on lookups" - you mean, "even running in check mode, your lookup plugin is going to do whatever it would do on a real run, so be careful . . . " That is a very important point - I'd make it more explicit here and above.

Reply:
Yes exactly. I may move some of this stuff about writes and implicit writes into its own document at some point as well, because while it can be trickier with lookups, some of it applies to modules too. 

For example: `vault_login` (the module) does not perform a login in check mode. `vault_read` (the module), takes all the same authentication parameters as `vault_login`, and will perform the login even in check mode, so that it can execute the read in check mode as well. The `vault_token_create` module (#213) similarly will perform the login with its authentication parameters, but will not create a token in check mode.

But for now I think I will like to keep some of this here, there's always room to reorganize and move things to new pages later.